internal/db/user_reference.go

```diff
@@ -0,0 +1,95 @@
+package db
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/edulinq/autograder/internal/model"
+)
+
+// Process a list of user inputs in the context of a course.
+// See model.CourseUserReferenceInput for the list of acceptable inputs.
+// Inputs are parsed without checking permissions.
+// APIs using this function should require a core.MinCourseRoleGrader or custom permissions checking.
+// Returns a reference, user errors, and a system error.
+// User-level errors return (partial reference, user errors, nil).
+// A system-level error returns (nil, nil, error).
+func ParseCourseUserReference(course *model.Course, rawReferences []model.CourseUserReferenceInput) (*model.CourseUserReference, map[string]error, error) {
```

Comment:
I don't see too much of a purpose for these soft errors.
I know you were just in the swamps of the job manager where we always needed soft and hard errors,
but here all likely errors are already the caller's fault.

I think you should have two versions of this function:
 - One that queries the db for the course users and then call the next function.
 - One that resolves the references given a list of course users.

With the above pattern, `courses/users/list` can now keep `CourseUsers`
(let the API infrastructure fetch those),
and then any errors that happen during parsing references is the users fault
(which means we can give an http 400 instead of 500).

Reply:
The second function may even be able to live in `model`.
(Code in lower dependency levels is generally preferable, since it doesn't have to "know" about a lot of other things.)

Reply:
Actually, another one of my comments calls into question the need for `courseUsers`.
So, you may not even need them or the first function.